experimental/dds/stack/src/sharedStack.ts

```diff
@@ -0,0 +1,162 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IEvent } from "@fluidframework/common-definitions";
+import { IsoBuffer } from "@fluidframework/common-utils";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { IChannel, IChannelAttributes, IChannelFactory, IChannelServices, IChannelStorageService, IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ITelemetryContext, ISummaryTreeWithStats, IGarbageCollectionData } from "@fluidframework/runtime-definitions";
+import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import { IFluidSerializer, SharedObject } from "@fluidframework/shared-object-base";
+
+/**
+ * Events useful for testing and debugging a {@link SharedStack}
+ */
+export interface DiagnosticEvents extends IEvent {
+    (event: "downloadedBlob", handler: () => void);
+    (event: "uploadedBlob", handler: () => void);
+    (event: "processedOp", handler: () => void);
+}
+
+/**
+ * A stack DDS which stores every element in its own attachment blob.
+ */
+export class SharedStack<T = unknown> extends SharedObject<DiagnosticEvents> {
+    private head?: IFluidHandle<ArrayBufferLike>;
+
+    /** Construct a new SharedStack */
+    public constructor(
+        id: string,
+        runtime: IFluidDataStoreRuntime,
+        attributes: IChannelAttributes,
+        telemetryContextPrefix: string
+    ) {
+        super(id, runtime, attributes, telemetryContextPrefix);
+    }
+
+    /** True iff there are no elements pushed to this stack */
+    public isEmpty(): boolean {
+        return this.head === undefined;
+    }
+
+    /** Download, remove and return the top element on this stack */
+    public async pop(): Promise<T> {
+        if (this.head === undefined) {
+            throw new Error("SharedStack is empty");
+        }
+
+        const head = await get<Node<T>>(this.head, this.serializer);
+        this.emit("downloadedBlob");
+        this.setHead(head.next);
+        return head.content;
+    }
+
+    /** Upload a new element and add it to the top of the stack */
+    public async push(content: T): Promise<void> {
+        const head: Node<T> = {
+            content,
+            next: this.head
+        };
+
+        this.setHead(await this.uploadNode(head));
```

Comment:
If I call stack.push(slowToUpload) followed by stack.push(fastToUpload), is there anything guaranteeing that fastToUpload ends up on top of slowToUpload?

Reply:
Or alternatively, is it up to the developer to `await` the first push before calling any subsequent pushes, if the order matters (making this kind of a consensus-style DDS)?

Reply:
Right, by the same logic as the `pop` scenario, I'd say they should await the call to guarantee that behavior. The downside is that if they want to upload multiple things in a deterministic order but have the uploads happen in parallel, they can't. I think I'm okay with that though; I'm trying to keep the implementation of this DDS as simple as possible for now. I didn't even give it a `peek`! :)

What do you mean by a "consensus-style DDS"?

Reply:
We probably need a better term :-P.  I'm drawing a comparison to the ConsensusRegisterCollection, which disallows double-writes before seeing the ack of the first write by only accepting writes whose referenceSequenceNumber indicates they have seen the latest state.

If the only "correct" way to use a DDS is to await for it to enter a "safe" state, then we should probably make that "safe" state discoverable (i.e. canEdit()) and error for attempted calls that we think are bad.

Reply:
Got it! I think for the sake of keeping this DDS implementation simple, I'll try changing it to work that way. That will make the APIs less confusing as well, addressing your other comments.

Reply:
I realized that for a consensus-style approach, it would take two op round trips to be able to perform a push (one to wait for sequencing of the operation, and another to upload the next blob, send the handle and wait for sequencing again).

Reply:
PUSH does not require any ops, or at least no synchronous waiting, because result of the operation is void (or at least I assume so - I do not see much reason to make it a Promise - set() on a map does not return a promise, this is not much different, and in general we should move toward not async workflows when it comes to changing data model as they do not work with offline / container serialization / rehydration - fully solving that though required proper 3-way merge capabilities that we are lacking).

And one op is sufficient to fully complete it. You can upload blob first (with a reference to parent commit), and then do CAS op to make it the HEAD of commit chain. If it fails (CAS assumes success only if parent commit did not change), you start over (including uploading blob). If this is unacceptable, you would need to push responsibility to generate commit chain to dedicated client (summarizer), with main client only conveying (through the op) that new content was pushed on the stack (content is in the op or in the blob, but no CAS / references to parent commit).